Attempt at reducing generated derivative size in mortar contact

recuero opened the pull request:
- Make thermomechanical classes that work with Lagrange element only add residual/derivatives on lower-dimensional domains
- Eliminate interior node derivatives from displacements and velocities.
- Use `sparsity_trim` to allow the user to reduce the derivative size in case there are additional cases where meaningless derivatives (of value less than the machine epsilon hopefully) are being generated.

Closes #20488.

recuero committed:
Make ADMortarLagrangeConstraint act only on lower-d domains (#20488)

Before we made the distinction of whether the user wanted to interpolate normals.
This is was done to avoid modifying existing behavior. But internal node normals and
tangents have no real meaning in this context and may contribute to populating the AD
derivative container with spurious information.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/978368/) on 092dc57 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20500/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/978376/) on 092dc57 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>643a67</th>
      <th colspan=4>#20500 092dc5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.02%</td>
      <td>89.00%</td>
      <td>-0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2911</td>
      <td>2904</td>
      <td>-7</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>359</td>
      <td>359</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20500/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20500/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20500/coverage/xfem)

</details>



This comment will be updated on new commits.

recuero committed:
Trim ADReal derivative size for Lagrange elements in mortar (#20488)

There is a risk in trimming derivatives since derivatives with small
absolute values can potentially be meaningful and necessary for PJFNK
and NEWTON convergence. However, in practical problems, we have seen that
ADReal can sometimes have derivatives on the order 1.0e-23 ~ 1.0e-27, which
are well below the machine epsilon and turn out to be many orders of magnitude
lower than other derivative sets. Since accounting for derivatives isn't free,
this commit adds a handle for removing derivatives that may not be numerically
relevant. An option to change the threshold value is provided.

recuero committed:
Interior node AD derivative removal. (#20488)

When coupling AD variables into mortar constraints acting on Lagrange
elements, the interior nodes carry derivative information that may be
useless to the mortar contact problem. Since the number of derivatives
needed for a mortar problem can be large (e.g. 3 displacements + temperature
+ 1 LM for normal pressure + 2 LM for frictional pressure + 1 LM for heat flux),
a problem that is exarcebated by secondary-primary projections, the total number of
derivatives required can balloon (to the point of running out of stack).

Removing unnecessary interior node derivatives can cut in a third this
number. For example, for a HEX27 element, we'd only require derivative info
pertaining to 9 nodes, instead of 27.

lindsayad committed:
Add unconstrained process derivatives method

Refs #000

recuero committed:
Fix interior node derivative trimming (#20488)

Forgot to account for primary/secondary surface differences when refactoring.

recuero committed:
Call processUnconstrainedDerivatives and test the work (#20488)

Also put this PR to the test. If this all works well, a simple normal-contact-only
with first order three dimensional elements should run _without_ having to
increase the AD derivative size:

One node on the secondary surface depends on 9 nodes (four adjacent first order faces),
on the primary side there are four more nodes. We have three displacement variables, so
the final displacement-based constraint will have (9 + 4) nodes * 3 disps = 39 components.

lindsayad commented:
@recuero you are slaying coverage in the contact module. I love it!

lindsayad commented on framework/include/constraints/ADMortarConstraint.h:
```diff
@@ -36,6 +36,25 @@ class ADMortarConstraint : public MortarConstraintBase
+   */
+  void
+  trimInteriorNodeDerivatives(const std::map<unsigned int, unsigned int> & primary_ip_lowerd_map,
+                              std::array<MooseVariable *, 3> & moose_var,
```
There's kind of a vector physics assumption baked into this API. Could we change this to a `std::vector` in case someone ever wants to pass in a single nodal variable (or two or > 3)?

recuero replied:
Switched inputs to `std::vector`. Makes more sense now for general use.

lindsayad commented on framework/src/constraints/ADMortarConstraint.C:
```diff
@@ -170,3 +170,69 @@ ADMortarConstraint::computeJacobian(Moose::MortarType mortar_type)
+    ADReal & var3,
+    const bool is_secondary)
+{
+  // Remove interior node variable's derivatives from AD objects.
```
Can we assert that the moose variables are nodal somewhere in this method?

recuero replied:
Should be done.

recuero committed:
Address Alex's comments

lindsayad commented on framework/src/constraints/ADMortarConstraint.C:
```diff
@@ -199,40 +199,27 @@ ADMortarConstraint::trimDerivative(const dof_id_type & remove_derivative_index,
-        trimDerivative(remove_derivative_index_z, var3);
-      }
+      for (const auto moose_var_index : make_range(moose_vars.size()))
+        if (moose_vars[moose_var_index])
```
are there cases where we put `nullptrs` in this container?

recuero replied:
Here, for example, https://github.com/idaholab/moose/blob/239e52b718b348acd7376dfa8f9db26c7538ed81/modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C#L91. But I could do `push_back` operations in the contact constraints instead and avoid it.

recuero replied:
With the latest push we shouldn't need to check for `nullptr` anymore.

recuero commented:
@lindsayad can you invalidate the controlled app recipe when you come back here? I don't think it has to do with these changes.

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/985725/) on de1e6e1 : invalidated by @lindsayad

lindsayad committed:
Add framework coverage of trim*Derivative

Some API changes as well:
- use std::vector<ADReal> instead of std::vector<ADReal *>
- template trimInteriorNodeDerivatives to allow for multiple types
  of containers

I want to allow for multiple types of containers because I don't
want users having to do dynamic memory allocation in methods like
`computeQpProperties`. I feel like that could be slow. So that means
allowing `std::array` ... which lead me to undo a change I forced
Antonio to make. Now we allow the variable containers to include
nullptrs for logical LIBMESH_DIM variable sets such as displacements.

There was also a minor bug fix in this commit. The passed-in size
of the dual numbers container may be different from the size of the
variables container.

lindsayad committed:
Add missing z-component gap_vec derivative computation

lindsayad committed:
Remove hopefully redundant relative velocity derivative computation

recuero commented on framework/include/constraints/ADMortarConstraint.h:
```diff
@@ -73,3 +74,39 @@ class ADMortarConstraint : public MortarConstraintBase
+        const auto remove_derivative_index = is_secondary
+                                                 ? moose_var->dofIndices()[dof_index]
+                                                 : moose_var->dofIndicesNeighbor()[dof_index];
+        for (auto & dual_number : dual_numbers)
+          trimDerivative(remove_derivative_index, dual_number);
```
Now we look for all the variable indices in the dual numbers. I think this makes it more general.

recuero commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -107,28 +107,9 @@ ComputeDynamicFrictionalForceLMMechanicalContact::computeQpProperties()
-        (*_secondary_z_dot)[_qp].derivatives() - (*_primary_z_dot)[_qp].derivatives();
-
-  _qp_tangential_velocity_nodal = relative_velocity * (_JxW_msm[_qp] * _coord[_qp]);
+  // It appears that the relative velocity between weighted gap and this class have a sign
```
Next time I touch this, I'll look into making this more consistent and avoid this ugly negative sign here.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/988213/) on 1d8477b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20500/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20500/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83ee3d4c9d789f2dc71f5c15a45b37362250681c`

lindsayad commented on framework/include/constraints/ADMortarConstraint.h:
```diff
@@ -36,6 +36,24 @@ class ADMortarConstraint : public MortarConstraintBase
+  /**
+   * Get rid of AD derivative entries by dof index
+   */
+  void trimDerivative(dof_id_type remove_derivative_index, ADReal & dual_number);
```
since it's pass by value, the caller of this method knows that whatever they pass from the callee will not be modified, so from an API standpoint that is enough. You'll note that in the *implementation* I mark the argument as `const`

lindsayad reviewed (approved):
Approved, assuming everything passes again. Since I have non-trivial commits on this now, maybe it would be good to get a third set of eyes on this? Maybe @dschwen ?

dschwen commented on framework/include/constraints/ADMortarConstraint.h:
```diff
@@ -56,3 +74,39 @@ class ADMortarConstraint : public MortarConstraintBase
+{
+  // Remove interior node variable's derivatives from AD objects.
+  for (const auto dof_index :
+       (is_secondary ? make_range(_test_secondary.size()) : make_range(_test_primary.size())))
```
Would be cool if index_range worked here

lindsayad commented:
@recuero please resolve and merge as you see fit
